04-frequent-scenarios.md

```diff
@@ -0,0 +1,288 @@
+
+The following cache control mechanisms are supported by the browser cache:
+* do not cache; also known as 'no-store'. This forbids your browser to cache at all.
+* use cache, but verify first; also known as 'no-cache' (don't ask). The browser will first fetch the resource from the server, passing one of two, or both optional headers (If-Modified-Since, If-None-Match). If the servers returns a 304 status code, it will use its cached resource. Otherwise, it will use the response and cache it.
+* use cache, but verify first if resource is stale; also known as 'must-revalidate'. Works similar to 'no-cache', but only calls the server if the maximum age has not been exceeded. Additionaly, on calling the server, the age will be updated if the resource is still fresh.
+
+For regular browser caches, the 'no-cache' option is preferred. Passing a max age, means it will be hard to evict a resource from cache until it has grown stale. The 'no-cache' has no such problems.
+
+A case can be made to use 'must-revalidate' in combination with a 'max-age', especially for sites that expect high traffic load on resources that are consumed manifold, but updated rarely.
+
+Request
+```
+GET /cars/1
+If-Modified-Since: [DATE]
+If-None-Match: [ETAG]
+```
+
+Response
+```
+Status code: 304
+Cache-Control: [CACHE INSTRUCTIONS]
+Last-Modified: [DATE] 
+Etag: [ETAG]
+```
+
+No special measures have to be made in the browser application. The browser cache will automatically take care of the response and all required actions to update the cache.
+
+The values for the Last-Modified and/or Etag are determined by the server. The client will use those values literally.
+
+If you want a controller methode that can deal with a request header containing, for example, If-Modified-Since, this is what it could look like:
+
+```java
+public static final String HTTP_DATE_FORMAT = "EEE, dd MMM yyyy HH:mm:ss zzz";
+public static final DateTimeFormatter HTTP_DATE_FORMATTER = DateTimeFormatter.ofPattern(HTTP_DATE_FORMAT);
+
+@RequestMapping(value = "/{id}", method = RequestMethod.GET)
+public Car getCar(
+        HttpServletResponse response,
+        @PathVariable String id,
+        @RequestHeader(value = "If-Modified-Since", required = false)
+            @DateTimeFormat(pattern = HTTP_DATE_FORMAT) LocalDateTime ifModifiedSince) {
+    Car car = carService.findCarIfStale(id, ifModifiedSince);
+    response.setHeader("Last-Modified", car.getModified().format(HTTP_DATE_FORMATTER));
+    return car;
+}
+```
+
+Ideally, findCarIfStale in the example above, would throw a not-modified exception if the resource is still fresh. A separate @ControllerAdvice method can then deal with this exception:
```

Comment:
I guess wether or not this particular situation calls for an exception is indeed debatable. There are other ways to deal with flows that contain more than one response path other than using [exceptions for control flow](http://programmers.stackexchange.com/questions/189222/are-exceptions-as-control-flow-considered-a-serious-antipattern-if-so-why). I do agree however that in some cases it can be pragmatic to throw an exception to let the ControllerAdvice handle it.

Right now you would also have to duplicate the logic whether or not an entity has changed in the meantime in all the different service methods (for cars, bikes, etc). That doesn't really make the code on that layer cleaner either in my opinion.

Maybe the problem could also be solved by implementing some custom filter that returns a status 304 if the record has not been modified by inspecting a caching layer of some sort before it gets into the controller. I don't have any first-hand experience with this but perhaps something like Varnish or Redis could also be used to move this kind of logic out of the server layer (and thus also prevent the necessity for an exception).